src/librustc_borrowck/borrowck/mod.rs

```diff
@@ -1135,6 +1093,102 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
         }
     }
 
+    fn note_and_explain_mutbl_error(&self, db: &mut DiagnosticBuilder, err: &BckError<'tcx>,
+                                    error_span: &Span) {
+        match err.cmt.note {
+            mc::NoteClosureEnv(upvar_id) | mc::NoteUpvarRef(upvar_id) => {
+                // If this is an `Fn` closure, it simply can't mutate upvars.
+                // If it's an `FnMut` closure, the original variable was declared immutable.
+                // We need to determine which is the case here.
+                let kind = match err.cmt.upvar().unwrap().cat {
+                    Categorization::Upvar(mc::Upvar { kind, .. }) => kind,
+                    _ => bug!()
+                };
+                if kind == ty::ClosureKind::Fn {
+                    db.span_help(self.tcx.map.span(upvar_id.closure_expr_id),
+                                 "consider changing this closure to take \
+                                 self by mutable reference");
+                }
+            }
+            _ => {
+                let mut cmt = &err.cmt;
+                while let Categorization::Deref(ref inner_cmt, ..) = cmt.cat {
+                    cmt = inner_cmt;
+                }
+                if let Categorization::Local(local_id) = cmt.cat {
+                    let parent = self.tcx.map.get_parent_node(local_id);
+                    let may_be_function = match self.tcx.map.get(parent) {
+                        hir::map::NodeItem(item) =>
+                            if let hir::ItemFn(ref function, ..) = item.node {
+                                Some(function)
+                            } else {
+                                None
+                            },
+                        hir::map::NodeImplItem(item) =>
+                            if let hir::ImplItemKind::Method(ref sig, ..) = item.node {
+                                Some(&sig.decl)
+                            } else {
+                                None
+                            },
+                        hir::map::NodeTraitItem(item) =>
+                            if let hir::MethodTraitItem(ref sig, ..) = item.node {
+                                Some(&sig.decl)
+                            } else {
+                                None
+                            },
+                        _ => None
+                    };
+
+                    if let Some(function) = may_be_function {
```

Comment:
Nit: variables of this type are usually called `fn_decl` or `decl`, not `function`, at least by me (since they are a `FnDecl`)